driver-sync/src/main/com/mongodb/client/internal/KeyManagementService.java

```diff
@@ -113,46 +111,55 @@ private void closeSocket(final Socket socket) {
         }
     }
 
-    private static class OperationTimeoutAwareInputStream extends InputStream {
+    private static final class OperationTimeoutAwareInputStream extends InputStream {
         private final Socket socket;
         private final Timeout operationTimeout;
         private final InputStream wrapped;
 
         /**
-         *
          * @param socket - socket to set timeout on.
          * @param operationTimeout - non-infinite timeout.
          */
-        OperationTimeoutAwareInputStream(final Socket socket, final Timeout operationTimeout) throws IOException {
-            assertFalse(operationTimeout.isInfinite());
+        private OperationTimeoutAwareInputStream(final Socket socket, final Timeout operationTimeout) throws IOException {
             this.socket = socket;
             this.operationTimeout = operationTimeout;
             this.wrapped = socket.getInputStream();
         }
 
-        private void setSoTimeout(final Socket socket, final Timeout operationTimeout) throws SocketException {
-            int remainingMs = (int) operationTimeout.remaining(MILLISECONDS);
-            if (operationTimeout.hasExpired()) {
-                throw TimeoutContext.createMongoTimeoutException("Timeout has expired while reading from KMS server");
-            }
-            socket.setSoTimeout(remainingMs);
+        public static InputStream wrapIfNeeded(@Nullable final Timeout operationTimeout, final SSLSocket socket) throws IOException {
+            return Timeout.checkedRun(operationTimeout, NANOSECONDS,
+                    () -> socket.getInputStream(),
+                    (ns) -> new OperationTimeoutAwareInputStream(socket, assertNotNull(operationTimeout)),
+                    () -> new OperationTimeoutAwareInputStream(socket, assertNotNull(operationTimeout)),
+                    () -> socket.getInputStream());
+        }
```

Comment:
Perhaps the underlying problem that you identify, of the first and last params, is solved by adding `nullAsInfinite`, which I have pushed up:

```
    return Timeout.nullAsInfinite(operationTimeout).checkedRun(NANOSECONDS,
                    () -> socket.getInputStream(),
                    (ns) -> new OperationTimeoutAwareInputStream(socket, assertNotNull(operationTimeout)),
                    () -> new OperationTimeoutAwareInputStream(socket, assertNotNull(operationTimeout)));
    }
```

This makes it explicit why the first and last are the same, and removes redundant code. It also facilitates extracting that wrapping, which I think belongs [outside of the method](https://github.com/mongodb/mongo-java-driver/pull/1215/files#r1537863409). It also means we can get rid of static variants. 

I found it helpful when implementing to think of params as going from "least expired to most expired" (infinite never expires, then remaining, then expired). I am not sure how helpful this is for others, but I do not recall needing to check param order when making changes. Perhaps this "order mnemonic" addresses remaining readability issues? (Early on, @stIncMale and I briefly discussed the fluent approach. I think it would take more work and design, and would have its own downsides.)

Reply:
I agree with @vbabanin using named handlers over unnamed parameters makes the code self explaining / readable. That said this pattern is similar to how a reactive stream {{Subscriber}} has named events, but then libraries (Reactor) often add helpers eg: `subscribe(Subscriber)` gets extended out to `subscribe(OnNext), subscribe(OnNext, onError), subscribe(OnNext, OnError, OnComplete)` .  Once you start using it and get familiar with it you often use the convience methods. 🤷

Reply:
I understand the rationale behind the "least expired to most expired" parameter organization. However, concerns arise regarding its long-term clarity and potential for error as this implicit understanding might not be as apparent in the future.

The approach with fluent API, while initially demanding more effort, promises greater maintainability in the future, making it easier to understand, modify, and debug.

Nevertheless, I respect the current direction chosen by the team and am willing to align with the consensus.

Seeing that @stIncMale and @ross are comfortable with the current implementation and considering the existing approvals on the PR, I believe you can go ahead and merge the PR without needing my explicit approval.

Reply:
> @ross

Probably not the Ross you're looking for.